specification/cognitiveservices/resource-manager/Microsoft.CognitiveServices/stable/2021-04-30/cognitiveservices.json

```diff
@@ -0,0 +1,2071 @@
+          "enum": [
+            "Accepted",
+            "Creating",
+            "Deleting",
+            "Moving",
+            "Failed",
+            "Succeeded",
+            "ResolvingDNS"
+          ],
+          "x-ms-enum": {
+            "name": "ProvisioningState",
+            "modelAsString": true
+          }
+        },
+        "endpoint": {
+          "type": "string",
+          "description": "Endpoint of the created account.",
+          "readOnly": true
+        },
+        "capabilities": {
+          "readOnly": true,
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/SkuCapability"
+          },
+          "description": "Gets the capabilities of the cognitive services account. Each item indicates the capability of a specific feature. The values are read-only and for reference only."
+        },
+        "isMigrated": {
+          "type": "boolean",
+          "description": "If the resource is migrated from an existing key.",
+          "readOnly": true
+        },
+        "migrationToken": {
+          "type": "string",
+          "description": "Resource migration token."
+        },
+        "skuChangeInfo": {
+          "$ref": "#/definitions/SkuChangeInfo",
+          "description": "Sku change info of account.",
+          "readOnly": true
+        },
+        "customSubDomainName": {
+          "type": "string",
+          "description": "Optional subdomain name used for token-based authentication."
+        },
+        "networkAcls": {
+          "$ref": "#/definitions/NetworkRuleSet",
+          "description": "A collection of rules governing the accessibility from specific network locations."
+        },
+        "encryption": {
+          "$ref": "#/definitions/Encryption",
+          "description": "The encryption properties for this resource."
+        },
+        "userOwnedStorage": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/UserOwnedStorage"
+          },
+          "description": "The storage accounts for this resource."
+        },
+        "privateEndpointConnections": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/PrivateEndpointConnection"
+          },
+          "description": "The private endpoint connection associated with the Cognitive Services account."
+        },
+        "publicNetworkAccess": {
+          "type": "string",
+          "description": "Whether or not public endpoint access is allowed for this account. Value is optional but if passed in, must be 'Enabled' or 'Disabled'",
+          "enum": [
+            "Enabled",
+            "Disabled"
+          ],
+          "x-ms-enum": {
+            "name": "PublicNetworkAccess",
+            "modelAsString": true
+          }
+        },
+        "apiProperties": {
+          "type": "object",
+          "additionalProperties": {
```

Comment:
>additionalProperties [](http://example.com/codeflow?start=11&length=20)

ARM recommends against the use of this pattern. Open-ended unschematized objects are problematic, even though they appear to make some things easier for RP owners. ARM is evolving toward stronger typing wherever possible because it allows all consumers of your API to reason about it. This has benefits for everyone that interacts with your API from infrastructure, to other Azure features to SDK owners to customers. Many historical APIs use this, but it's a pattern that should be used only as a last resort. 

For example, customers will not be able to manage these objects using Azure Policy, which is a standard feature. Your customers may be unhappy about that. There are other features like ARG that won't be able to support these objects, and others features and capabilities in the pipeline.

To address this you should enumerate the object properties in your swagger. If that isn't a good enough option, you could implement this object as a resource type. This is recommended, but it is ultimately your decision.

Reply:
We would really like to enumerate all object properties if we can.

However Cognitive Services is in fact a collection of services, some backend services need specific extra metadata, and this `apiProperties` is that metadata.

Our Resource Provider doesn't use this metadata, in fact RP doesn't know what properties are available for specific backend services. So we think an unschematized objects is best option in here.